File: packages/hqtui/src/buffer.ts
import { DEFAULT_COLOR, type Color } from "./color.ts";
import {
  CONTINUATION, REPLACEMENT, cellWidth, graphemes, cellText, isLoneSurrogate, isUnsafeCodepoint,
} from "./unicode.ts";

/** Style attribute bits packed into the cell's 16-bit attribute slot. */
export const Attr = {
  None: 0,
  Bold: 1 << 0,
  Dim: 1 << 1,
  Italic: 1 << 2,
  Underline: 1 << 3,
  Blink: 1 << 4,
  Reverse: 1 << 5,
  Strike: 1 << 6,
} as const;

export type Attributes = number;

export interface Style {
  fg?: Color;
  bg?: Color;
  attrs?: Attributes;
}

/**
 * The screen is one grid of cells, not a tree of widgets. Four parallel typed
 * arrays keep a frame allocation-free: no object is created per cell, ever.
 */
export class FrameBuffer {
  width: number;
  height: number;
  chars: Uint32Array;
  fg: Uint32Array;
  bg: Uint32Array;
  attrs: Uint16Array;

  constructor(width: number, height: number) {
    this.width = Math.max(0, width | 0);
    this.height = Math.max(0, height | 0);
    const n = this.width * this.height;
    this.chars = new Uint32Array(n);
    this.fg = new Uint32Array(n);
    this.bg = new Uint32Array(n);
    this.attrs = new Uint16Array(n);
    this.clear();
  }

  /** Resize in place, reusing the existing allocation when it is large enough. */
  resize(width: number, height: number): void {
    const w = Math.max(0, width | 0);
    const h = Math.max(0, height | 0);
    if (w === this.width && h === this.height) return;
    const n = w * h;
    if (n > this.chars.length) {
      this.chars = new Uint32Array(n);
      this.fg = new Uint32Array(n);
      this.bg = new Uint32Array(n);
      this.attrs = new Uint16Array(n);
    }
    this.width = w;
    this.height = h;
    this.clear();
  }

  index(x: number, y: number): number {
    return y * this.width + x;
  }

  clear(fill: Color = DEFAULT_COLOR, fg: Color = DEFAULT_COLOR): void {
    const n = this.width * this.height;
    this.chars.fill(32, 0, n);
    this.fg.fill(fg, 0, n);
    this.bg.fill(fill, 0, n);
    this.attrs.fill(0, 0, n);
  }

  inBounds(x: number, y: number): boolean {
    return x >= 0 && y >= 0 && x < this.width && y < this.height;
  }

  /** Write one already-decoded cell value. Returns columns consumed. */
  setCell(x: number, y: number, value: number, style?: Style): number {
    if (!this.inBounds(x, y)) return 0;
    // The last line of defence: this is the only path that writes a character
    // into the grid, and the encoder hands cell text straight to the terminal.
    // Refusing unsafe values here means the buffer *cannot* hold a live escape,
    // whatever the caller passes — including the low-level escape hatch.
    //
    // Validate what will actually be stored, not what was passed. `chars` is a
    // Uint32Array, so it applies ToUint32: 2**32 + 0x1b looks like a harmless
    // large number but truncates to a live ESC. Coerce first, then check.
    // A lead-less continuation is not writable either; it would silently eat a
    // column out of the row. Ordered so printable ASCII costs one comparison.
    value = value >>> 0;
    if (value >= 0x7f) {
      if (isUnsafeCodepoint(value) || value === CONTINUATION) value = 32;
      // A lone surrogate would fuse with its neighbour into a single glyph.
      else if (isLoneSurrogate(value)) value = REPLACEMENT;
    } else if (value < 0x20) {
      value = 32;
    }
    const w = cellWidth(value);
    const i = this.index(x, y);
    // Overwriting the tail of a wide char to our left would orphan it; blank it.
    if (this.chars[i] === CONTINUATION && x > 0) this.chars[i - 1] = 32;
    this.chars[i] = value;
    if (style) {
      if (style.fg !== undefined) this.fg[i] = style.fg;
      if (style.bg !== undefined) this.bg[i] = style.bg;
      if (style.attrs !== undefined) this.attrs[i] = style.attrs;
    }
    if (w === 2) {
      if (x + 1 < this.width) {
        const j = i + 1;
        this.chars[j] = CONTINUATION;
        if (style) {
          if (style.fg !== undefined) this.fg[j] = style.fg;
          if (style.bg !== undefined) this.bg[j] = style.bg;
          if (style.attrs !== undefined) this.attrs[j] = style.attrs;
        }
      } else {
        // No room for the second half: draw a space rather than corrupt the row.
        this.chars[i] = 32;
        return 1;
      }
    }
    return Math.max(1, w);
  }

  /** Write text left to right. Returns the number of columns written. */
  write(x: number, y: number, text: string, style?: Style, maxWidth = Infinity): number {
    if (y < 0 || y >= this.height) return 0;
    let cx = x;
    let used = 0;
    for (const g of graphemes(text)) {
      if (used + g.width > maxWidth) break;
      if (cx >= this.width) break;
      if (cx + g.width > this.width) break;
      if (cx >= 0) {
        this.setCell(cx, y, g.value, style);
      }
      cx += g.width;
      used += g.width;
    }
    return used;
  }

  fillRect(x: number, y: number, w: number, h: number, ch = 32, style?: Style): void {
    const x0 = Math.max(0, x);
    const y0 = Math.max(0, y);
    const x1 = Math.min(this.width, x + w);
    const y1 = Math.min(this.height, y + h);
    for (let cy = y0; cy < y1; cy++) {
      for (let cx = x0; cx < x1; cx++) this.setCell(cx, cy, ch, style);
    }
  }

  /** Restyle a region without touching its characters. */
  styleRect(x: number, y: number, w: number, h: number, style: Style): void {
    const x0 = Math.max(0, x);
    const y0 = Math.max(0, y);
    const x1 = Math.min(this.width, x + w);
    const y1 = Math.min(this.height, y + h);
    for (let cy = y0; cy < y1; cy++) {
      for (let cx = x0; cx < x1; cx++) {
        const i = this.index(cx, cy);
        if (style.fg !== undefined) this.fg[i] = style.fg;
        if (style.bg !== undefined) this.bg[i] = style.bg;
        if (style.attrs !== undefined) this.attrs[i] = style.attrs;
      }
    }
  }

  /** Copy another buffer's contents into this one (same dimensions assumed). */
  copyFrom(other: FrameBuffer): void {
    const n = Math.min(this.width * this.height, other.width * other.height);
    this.chars.set(other.chars.subarray(0, n));
    this.fg.set(other.fg.subarray(0, n));
    this.bg.set(other.bg.subarray(0, n));
    this.attrs.set(other.attrs.subarray(0, n));
  }

  /** Plain text of one row, for tests and headless rendering. */
  rowText(y: number): string {
    if (y < 0 || y >= this.height) return "";
    let out = "";
    for (let x = 0; x < this.width; x++) {
      const v = this.chars[this.index(x, y)];
      if (v === CONTINUATION) continue;
      out += v === 0 ? " " : cellText(v);
    }
    return out;
  }

  /** Whole buffer as plain text. */
  toText(trimEnd = true): string {
    const rows: string[] = [];
    for (let y = 0; y < this.height; y++) {
      const r = this.rowText(y);
      rows.push(trimEnd ? r.replace(/\s+$/, "") : r);
    }
    return rows.join("\n");
  }
}

File: ports/python/hqtui/buffer.py
"""The screen is one grid of cells, not a tree of widgets.

Four parallel arrays keep a frame allocation-free: no object is created per
cell, ever. Python's ``array`` module gives the same flat typed storage the
reference gets from ``Uint32Array``, and matters more here than there — a list
of boxed ints for a 200x50 screen is ten thousand objects per frame.
"""

from __future__ import annotations

from array import array
from dataclasses import dataclass
from typing import Iterable

from .color import Color, DEFAULT_COLOR
from .unicode import (
    CLUSTER_BASE,
    CONTINUATION,
    REPLACEMENT,
    cell_text,
    cell_width,
    graphemes,
    is_unsafe_codepoint,
)

__all__ = ["Attrs", "Style", "FrameBuffer"]


class Attrs:
    """Style attribute bits packed into a cell's 16-bit slot.

    A namespace of ints rather than an ``IntFlag``: these are combined and
    tested millions of times per frame, and ``IntFlag`` arithmetic allocates.
    """

    NONE = 0
    BOLD = 1 << 0
    DIM = 1 << 1
    ITALIC = 1 << 2
    UNDERLINE = 1 << 3
    BLINK = 1 << 4
    REVERSE = 1 << 5
    STRIKE = 1 << 6


@dataclass(frozen=True, slots=True)
class Style:
    """A partial style.

    ``None`` on a field means "leave whatever is already there", which is what
    lets a widget set only the foreground of a region.
    """

    fg: Color | None = None
    bg: Color | None = None
    attrs: int | None = None

    def merged(self, other: "Style") -> "Style":
        """Fill in anything this style leaves unset from ``other``."""
        return Style(
            fg=self.fg if self.fg is not None else other.fg,
            bg=self.bg if self.bg is not None else other.bg,
            attrs=self.attrs if self.attrs is not None else other.attrs,
        )


NO_STYLE = Style()

_INFINITY = 1 << 30


class FrameBuffer:
    """A grid of styled cells. The only thing the encoder ever reads."""

    __slots__ = ("width", "height", "chars", "fg", "bg", "attrs")

    def __init__(self, width: int, height: int) -> None:
        self.width = max(0, int(width))
        self.height = max(0, int(height))
        n = self.width * self.height
        self.chars = array("L", [32]) * n if n else array("L")
        self.fg = array("L", [0]) * n if n else array("L")
        self.bg = array("L", [0]) * n if n else array("L")
        self.attrs = array("H", [0]) * n if n else array("H")

    def resize(self, width: int, height: int) -> None:
        """Resize, reusing the existing allocation when it is large enough."""
        w, h = max(0, int(width)), max(0, int(height))
        if w == self.width and h == self.height:
            return
        n = w * h
        if n > len(self.chars):
            self.chars = array("L", [32]) * n if n else array("L")
            self.fg = array("L", [0]) * n if n else array("L")
            self.bg = array("L", [0]) * n if n else array("L")
            self.attrs = array("H", [0]) * n if n else array("H")
        self.width, self.height = w, h
        self.clear()

    def index(self, x: int, y: int) -> int:
        return y * self.width + x

    def clear(self, bg: Color = DEFAULT_COLOR, fg: Color = DEFAULT_COLOR) -> None:
        n = self.width * self.height
        self.chars[:n] = array("L", [32]) * n if n else array("L")
        self.fg[:n] = array("L", [int(fg)]) * n if n else array("L")
        self.bg[:n] = array("L", [int(bg)]) * n if n else array("L")
        self.attrs[:n] = array("H", [0]) * n if n else array("H")

    def in_bounds(self, x: int, y: int) -> bool:
        return 0 <= x < self.width and 0 <= y < self.height

    def _apply(self, i: int, style: Style) -> None:
        if style.fg is not None:
            self.fg[i] = int(style.fg)
        if style.bg is not None:
            self.bg[i] = int(style.bg)
        if style.attrs is not None:
            self.attrs[i] = style.attrs

    def set_cell(self, x: int, y: int, value: int, style: Style = NO_STYLE) -> int:
        """Write one already-decoded cell value. Returns columns consumed.

        This is the only path that writes a character into the grid, and the
        encoder hands cell text straight to the terminal. Refusing unsafe values
        here means the buffer *cannot* hold a live escape, whatever the caller
        passes — including the low-level escape hatch.
        """
        if not self.in_bounds(x, y):
            return 0

        # Validate what will actually be stored: the arrays are unsigned, so a
        # huge value truncates and could otherwise become a live ESC.
        value &= 0xFFFFFFFF
        # Ordered so printable ASCII costs one comparison. A lead-less
        # continuation is not writable either; it would silently eat a column.
        if value >= 0x7F:
            if is_unsafe_codepoint(value) or value == CONTINUATION:
                value = 32
            elif not _valid_scalar(value):
                value = REPLACEMENT
        elif value < 0x20:
            value = 32

        w = cell_width(value)
        i = self.index(x, y)
        # Overwriting the tail of a wide char to our left would orphan it.
        if self.chars[i] == CONTINUATION and x > 0:
            self.chars[i - 1] = 32
        self.chars[i] = value
        self._apply(i, style)

        if w == 2:
            if x + 1 < self.width:
                self.chars[i + 1] = CONTINUATION
                self._apply(i + 1, style)
            else:
                # No room for the second half: draw a space rather than corrupt
                # the row.
                self.chars[i] = 32
                return 1
        return max(1, w)

    def write(
        self, x: int, y: int, text: str, style: Style = NO_STYLE, max_width: int = _INFINITY
    ) -> int:
        """Write text left to right. Returns the number of columns written."""
        if y < 0 or y >= self.height:
            return 0
        cx, used = x, 0
        for g in graphemes(text):
            if used + g.width > max_width:
                break
            if cx >= self.width or cx + g.width > self.width:
                break
            if cx >= 0:
                self.set_cell(cx, y, g.value, style)
            cx += g.width
            used += g.width
        return used

    def fill_rect(
        self, x: int, y: int, w: int, h: int, ch: int = 32, style: Style = NO_STYLE
    ) -> None:
        x0, y0 = max(0, x), max(0, y)
        x1, y1 = min(self.width, x + w), min(self.height, y + h)
        for cy in range(y0, y1):
            for cx in range(x0, x1):
                self.set_cell(cx, cy, ch, style)

    def style_rect(self, x: int, y: int, w: int, h: int, style: Style) -> None:
        """Restyle a region without touching its characters."""
        x0, y0 = max(0, x), max(0, y)
        x1, y1 = min(self.width, x + w), min(self.height, y + h)
        for cy in range(y0, y1):
            row = cy * self.width
            for cx in range(x0, x1):
                self._apply(row + cx, style)

    def copy_from(self, other: "FrameBuffer") -> None:
        """Copy another buffer's contents (same dimensions assumed)."""
        n = min(self.width * self.height, other.width * other.height)
        self.chars[:n] = other.chars[:n]
        self.fg[:n] = other.fg[:n]
        self.bg[:n] = other.bg[:n]
        self.attrs[:n] = other.attrs[:n]

    def row_text(self, y: int) -> str:
        """Plain text of one row, for tests and headless rendering."""
        if y < 0 or y >= self.height:
            return ""
        row = y * self.width
        parts: list[str] = []
        for x in range(self.width):
            v = self.chars[row + x]
            if v == CONTINUATION:
                continue
            parts.append(" " if v == 0 else cell_text(v))
        return "".join(parts)

    def to_text(self) -> str:
        """Whole buffer as plain text, trailing whitespace trimmed per row."""
        from .unicode import _trim_end

        return "\n".join(_trim_end(self.row_text(y)) for y in range(self.height))

    def cells(self) -> Iterable[int]:
        return self.chars[: self.width * self.height]


def _valid_scalar(value: int) -> bool:
    """A cell value only holds a real character if it is a Unicode scalar.

    The reference reaches this check via lone surrogates, which a Python ``str``
    cannot contain; a caller using the raw ``set_cell`` escape hatch still can,
    so the guard stays.
    """
    if value >= CLUSTER_BASE:
        return True
    return value <= 0x10FFFF and not (0xD800 <= value <= 0xDFFF)
